src/transformStateWithClones.js

```diff
@@ -7,7 +7,28 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    currentState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      for (const key in action.extraData) {
+        currentState[key] = action.extraData[key];
+      }
+    } else if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    } else if (action.type === 'clear') {
+      currentState = {};
+    }
```

Comment:
Consider using a `switch` statement instead of multiple `if`/`else if` statements for better readability and to follow the checklist item regarding the use of `switch` when you have a limited amount of conditions.